src/Mvc/Mvc.Core/ref/Microsoft.AspNetCore.Mvc.Core.netcoreapp3.0.cs

```diff
@@ -2909,6 +2910,11 @@ public partial class ValidatorCache
 }
 namespace Microsoft.AspNetCore.Mvc.Routing
 {
+    public abstract partial class DynamicRouteValueTransformer
+    {
+        protected DynamicRouteValueTransformer() { }
+        public abstract System.Threading.Tasks.Task<Microsoft.AspNetCore.Routing.RouteValueDictionary> TransformAsync(Microsoft.AspNetCore.Http.HttpContext httpContext, Microsoft.AspNetCore.Routing.RouteValueDictionary values);
```

Comment:
ValueTask

Reply:
Why? Is the guidance to just use ValueTask **everywhere** now?

Reply:
I think the guidance (as of "Stephen Toub says so") is that if the method is mostly expected to return synchronously, use a ValueTask, but if the method is expected to do real async work (like let's say > 20% of the time, but I'm not sure where I read it, but that's somehow the number in my head) use Task, as ValueTask comes with it's own struct-based overhead and just wraps a real Task in that case..

Reply:
`ValueTask<T>`  is still the guidance, there's almost no reason to use `Task<T>` as it forces an allocation. You can also now reuse allocations with `ValueTask<T>`  so it's an extensibility point as well.

> Why? Is the guidance to just use ValueTask everywhere now?

Yes, generic `ValueTask<T>` absolutely, for `ValueTask` is more questionable and has more overhead and is only really valuable as an extensibility point (using `IValueTaskSource`)

Reply:
hrm didn't sound like it here: https://devblogs.microsoft.com/dotnet/understanding-the-whys-whats-and-whens-of-valuetask/#user-content-should-every-new-asynchronous-api-return-valuetask--valuetasktresult

" There are also some minor costs associated with returning a ValueTask<TResult> instead of a Task<TResult>, e.g. in microbenchmarks it’s a bit faster to await a Task<TResult> than it is to await a ValueTask<TResult>, so if you can use cached tasks (e.g. you’re API returns Task or Task<bool>), you might be better off performance-wise sticking with Task and Task<bool>. ValueTask / ValueTask<TResult> are also multiple words in size, and so when these are awaitd and a field for them is stored in a calling async method’s state machine, they’ll take up a little more space in that state machine object." an yadda.. tbh it doesn#t feel like a complete replacement for Task everywhere.

Reply:
> However, ValueTask / ValueTask<TResult> are great choices when a) you expect consumers of your API to only await them directly, b) allocation-related overhead is important to avoid for your API, and c) either you expect synchronous completion to be a very common case, or you’re able to effectively pool objects for use with asynchronous completion. When adding abstract, virtual, or interface methods, you also need to consider whether these situations will exist for overrides/implementations of that method.

If you have a virtual member, you can't assume what derived implementations will do. To optimize for common synchronous cases and offer extensibility for derived types, use `ValueTask<T>`.

The framework is the one that is awaiting this so it's mostly up to the implementation to just return synchronously or asynchronously. The other downsides don't apply.

I really want to go back the change a bunch of APIs to return `ValueTask<T>` 😄 . I could squash so many needless allocations in the auth stack..

Reply:
Appreciate the time you took to answer that. Hopefully someone finds this hidden conversation, reads it and learns, too.